### ISSUE TITLE
Clarification on concurrency with Puma + Ruby MRI

### ISSUE THREAD

**Issue by alexbrahastoll** [puma/puma#1003]
Hi,

First of all, I would like to say thank you to the Puma community for such a great open source project!

I was not able to find any directions in the README on the best channel for making a question about Puma, so please let me know if I should use a different channel.

I read the README sections about concurrency, the thread pool and the clustered mode (and also googled about Puma related discussions on these topics), but I still have a doubt that is bogging me: what is the difference between having multiple Puma processes (each with its own thread pool) versus having just one Puma process with a higher number of threads in its pool (i.e. the same total of threads as the clustered setup)? Are there any advantages / disadvantages for these approaches given we are using Ruby MRI?

Thanks a lot!

**Comment by evanphx** [comment]
Hi @alexbrahastoll,

I'm happy to clear that up and I also need to do a better job explaining it in the README!

Basically, because MRI's threads do not run in parallel (but do run concurrently) the max throughput of a process is capped by how much IO and such the app does. The reason for that is waiting on IO (talking to a DB, etc) will allow another thread to run.

To bump the throughput up, you need to add additional workers. The workers DO run in parallel but obviously consume additional memory to do so.

You'll have to do some experimenting with your app to find the right balance, but a good rule of thumb is run as many workers as you have cores in the computer (so long as you have enough memory that is). Then ratchet the number of threads up. If you're coming from Unicorn, I suggest people start with half the number of processes used in Unicorn and 2 threads per worker. That will save you half as much memory as Unicorn and likely your throughput will be the same.

Once you're happy with that, begin to experiment until you find a throughput to memory ratio that works for you.

**Comment by alexbrahastoll** [comment]
Thanks a lot for the clarification, @evanphx ! I was incorrectly assuming that MRI's GIL would prevent  serving requests in parallel even with multiple Puma processes. Good to know it is possible to scale by increasing the number of cores in the server (as long as enough memory is available for the new processes, as you explained) =)

**Comment by TengZhangAutoGrid** [comment]
@alexbrahastoll the reason why Puma is powerful is it's multi-threading feature. Multi-processing can be done by other servers like Unicorn. I think Puma is trying to solve the problem of **scale by increasing the number of cores in the server** by threading.

**Comment by TengZhangAutoGrid** [comment]
Hi @evanphx , I came here randomly. But, I feel you are the guy who can answer my questions. 
Recently, we started a new project using puma. According to our need, we have to make it support multi-tenancy. But, again, we have to make sure it's thread safe.

Here's a old solution: http://railscasts.com/episodes/388-multitenancy-with-scopes
But, I think it's not thread safe because it's using class varible. (Correct me if I'm wrong.)

Here's a gem called **Apartment** doing this: https://github.com/influitive/apartment
But, after some research on this, I still can't figure out if it's really thread safe.

Question1. Is **Apartment** thread safe with puma?
Question2. If it's not, is there any a good way or gem doing multi-tenancy, and compatible with Puma?
